File: detect_cycle/detectCycle.py
from typing import *
# Detect a cycle in a linked list

class Node:
    def __init__(self, new_data: int):
        self.data = new_data
        self.next = None

class Solution:
    def hasCycle(self, head: Node) -> Node:
        node = set()

        while head is not None:
            if head in node:
                return True
            node.add(head)
            head = head.next
        
        return False
# Test the function
solution = Solution()
head = Node(1)
head.next = Node(2)
head.next.next = Node(3)
head.next.next.next = Node(4)
head.next.next.next.next = Node(5)
head.next.next.next.next.next = head.next

if solution.hasCycle(head):
    print("Cycle detected")
else:
    print("No cycle detected")

File: detect_cycle/detectCycle.go
package main

// Detect a cycle in a linked list

type Node struct {
	data int
	next *Node
}

func detectCycle(head *Node) bool {
	// create a set/map to store the node
	node := map[*Node]bool{}

	for head != nil {
		// if head in node
		if _, ok := node[head]; ok {
			return true
		}
		node[head] = true // add head to node
		head = head.next
	}
	return false
}

func main() {
	head := &Node{data: 1}
	head.next = &Node{data: 2}
	head.next.next = &Node{data: 3}
	head.next.next.next = &Node{data: 4}
	head.next.next.next.next = head.next

	if detectCycle(head) {
		println("Cycle detected")
	} else {
		println("Cycle not detected")
	}
}
